### DISCUSSION TITLE
How to upload a file to a collection using *forms.RecordUpsert.Submit()?

### DISCUSSION THREAD

**Discussion by Kwameoab** [pocketbase/pocketbase#1194] [Q&A] [answered]
I am trying to extend PocketBase by adding a route which will upload an image from a url to a collection after modifying the image. 

I am struggling because I do not know what the expected value for a file type is in a requestData map. I have tried different value types such as base64String, []bytes, and *bytes.Reader to no avail. 

Am I doing something wrong or is it not possible to upload a file using this method? If it is not possible, what is the recommended way to upload a file to a collection when extending PocketBase. 

Thank you so much for taking time out of your day to help me with this. Below is the skeleton of my code:  

``` go
urlImageResp, _ := http.Get("https://example.com/")
urlImage, _, _ := image.Decode(urlImageResp.Body)

imageRequestData := make(map[string]any)
imageRequestData["image_content"] = ??? // <- What is the valid key here

imageCollection, _ := app.Dao().FindCollectionByNameOrId("all_images")
newImageRecord := models.NewRecord(imageCollection)
imageUpsert := forms.NewRecordUpsert(app, newImageRecord)

imageUpsert.LoadData(imageRequestData)
if err := imageUpsert.Submit(); err != nil {
		return err
}
```

**Comment by ganigeorgiev** [comment, marked as answer]
Unfortunately `form.RecordUpsert` currently supports loading files only if they are part of a multipart `http.Request`.

I've added a task to the [roadmap](https://github.com/orgs/pocketbase/projects/2) and support/helpers for binding remote/local files will be implemented in the next few days after https://github.com/pocketbase/pocketbase/issues/1187.

One workaround could be to construct a dummy multipart `http.Request` instance and load it with `LoadRequest(r, "")`, eg. something like (_I haven't tested it_):

```go
func createMultipartRequest(data map[string]string, files map[string]string) (*http.Request, error) {
	body := new(bytes.Buffer)
	mp := multipart.NewWriter(body)
	defer mp.Close()

	// write data fields
	for k, v := range data {
		mp.WriteField(k, v)
	}

	for fieldName, path := range files {
		w, err := mp.CreateFormFile(fieldName, filepath.Base(path))
		if err != nil {
			return nil, err
		}
		f, err := os.Open(path)
		if err != nil {
			return nil, err
		}
		if _, err := io.Copy(w, f); err != nil {
			f.Close()
			return nil, err
		}
		f.Close()
	}

	req, _ := http.NewRequest("", "", body)
	req.Header.Set("Content-Type", mp.FormDataContentType())

	return req, nil
}


// create a dummy request instance and load it into the form
req, _ := createMultipartRequest(map[string]string{
	"title": "abc",
}, map[string]string{
	"yourFileFieldName": "/path/to/file",
})

imageUpsert.LoadRequest(req, "")

// notes:
// - the above expect the uploaded file(s) to be local, but you can adjust the helper if you want to change it and accept a url (the method for example could download the file upfront in a temp location)
// - the above also shows how to load regular data to the same multipart body, but if you don't need it you can pass `nil`;  you can combine imageUpsert.LoadRequest and imageUpsert.LoadData just fine
```

_edit: updated the example to show writing the file content into the multipart writer_

**Reply by Kwameoab** [reply]
Thank you so much for this. The provided code worked as intended. I would be interested to know how the bindings worked out.

**Reply by ganigeorgiev** [reply]
@Kwameoab Just to let you know that I've refactored the file handling and data load of the `RecordUpsert` form and added 2 new helpers and a new file resource abstraction to load and remove files more easily:
```go
file1, _ := filesystem.NewFileFromPath("/path/to/file1.txt")
file2, _ := filesystem.NewFileFromPath("/path/to/file2.txt")

// add file(s)
form.AddFiles("yourFileFieldName", file1, file2)

// remove already uploaded or newly attached file(s)
form.RemoveFiles("yourFileFieldName", "file1.txt", "file2.txt")

// note: for both methods the actual upload and deletion happens only on successful form submission
```

The changes are currently only in `master` and will be available with the next release.

**Reply by EduFdezSoy** [reply]
for my javascript friends:
```js
const file1 = $filesystem.fileFromUrl(url)
// or
const file2 = $filesystem.fileFromPath(path)

// and then you just
form.addFiles("yourFileFieldName", file1, file2);
```
And remember:
>  for both methods the actual upload and deletion happens only on successful form submission
